File: utils_ts/verify.ts
// @ts-nocheck
/* eslint-disable */
import { deterministicMarshal } from "./encode";
import { Envelope } from "../ts/message";
import { sha256 } from "@bsv/sdk/primitives/Hash";
import { verify as brc77Verify } from "@bsv/sdk/messages/SignedMessage";
import PrivateKey from "@bsv/sdk/primitives/PrivateKey";
import { Buffer } from "buffer";

// anyone private key 0..01 (BRC-77)
const anyoneKey = new PrivateKey(Buffer.alloc(32, 0).map((v, i) => (i === 31 ? 1 : 0)));

export function verifyEnvelope(env: Envelope): boolean {
  if (env.signatureAlgo !== "ECDSA_SECP256K1_SHA256") return false;
  // build sig input
  const clone: Envelope = { ...env, signature: Buffer.alloc(0), signatureAlgo: "" } as Envelope;
  const sigInput = deterministicMarshal(clone);
  const digest = sha256(sigInput);
  const ok = brc77Verify(digest, env.signature, anyoneKey);
  return ok;
} 

File: pkg/verify/verify.go
package verify

import (
	"errors"

	v1 "github.com/spycat55/keymaster_proto/gen/go"
	encodepkg "github.com/spycat55/keymaster_proto/pkg/encode"

	ec "github.com/bsv-blockchain/go-sdk/primitives/ec"
	hash "github.com/bsv-blockchain/go-sdk/primitives/hash"
	"google.golang.org/protobuf/proto"
)

// VerifyEnvelope returns nil if signature valid else error.
func VerifyEnvelope(env *v1.Envelope) error {
	if env.SignatureAlgo != "ECDSA_SECP256K1_SHA256" {
		return errors.New("unsupported signature_algo")
	}
	pubKeyBytes := env.GetHeader().GetFromPubkey()
	if len(pubKeyBytes) == 0 {
		return errors.New("missing from_pubkey")
	}
	pub, err := ec.ParsePubKey(pubKeyBytes)
	if err != nil {
		return err
	}
	// rebuild input
	clone := proto.Clone(env).(*v1.Envelope)
	clone.Signature = nil
	clone.SignatureAlgo = ""
	b, err := encodepkg.DeterministicMarshal(clone)
	if err != nil {
		return err
	}
	digest := hash.Sha256(b)

	sig, err := ec.FromDER(env.Signature)
	if err != nil {
		return err
	}
	if !sig.Verify(digest, pub) {
		return errors.New("signature invalid")
	}
	return nil
}
